### PR TITLE
New: UDP Syslog support and Instance Name

### PR DESCRIPTION
**Pull request by RobinDadswell**
#### Database Migration
NO

#### Description
Allows for UDP Syslogging and adds instance name so you can see what logs come from which instance when running multiple instances. Instance name is used in both syslog and within the tab; and requires Sonarr to be in the instance name somewhere

This is already in Radarr, Readarr, Lidarr, Prowlarr and Whisparr; commits have been loosely cherry-picked from there (sonarr-ified where required and updated to make a bit more logical sense)

Syslog is a config file only parameter set.

#### Todos
- [ ] Tests
- [ ] Wiki Updates

#### Issues Fixed or Closed by this PR

* #4778

### CODE REVIEW

File: src/Sonarr.Api.V3/Config/HostConfigModule.cs [hunk 1]
```diff
@@ -38,6 +38,7 @@ public HostConfigModule(IConfigFileProvider configFileProvider, IConfigService c
             SharedValidator.RuleFor(c => c.Port).ValidPort();
 
             SharedValidator.RuleFor(c => c.UrlBase).ValidUrlBase();
+            SharedValidator.RuleFor(c => c.InstanceName).ContainsSonarr().When(c => c.InstanceName.IsNotNullOrWhiteSpace());
```
Comment (markus101) [inline src/Sonarr.Api.V3/Config/HostConfigModule.cs:41]: This shouldn't be conditional on it not being null, otherwise the value can be set to nothing and the UI now shows nothing or `Page Title - `.

Removing correctly validates that the name isn't null/empty and contains Sonarr, which is the expected outcome.

**Review by markus101 — CHANGES_REQUESTED** [review]
Just the one issue with validation that needs to be resolved.

**Comment by Taloth** [comment]
Question regarding syslog. Why not leave it to journalctl to forward to syslog?

**Comment by RobinDadswell** [comment]
> Question regarding syslog. Why not leave it to journalctl to forward to syslog?

That's fine if being run on Linux, what about if its being run on Windows and wants syslogging. Or if it's on a seedbox without access to configure journalctl or rsyslog or similar and you want centralised logging back to your own syslog server etc.

**Comment by RobinDadswell** [comment]
This also allows for syslog to be trace but files to be info as an example

File: src/Sonarr.Api.V3/Config/HostConfigModule.cs (hunk 1 above)
Comment (RobinDadswell) [reply src/Sonarr.Api.V3/Config/HostConfigModule.cs]: Sorted, thanks for picking that one up!

**Comment by Taloth** [comment]
> > Isn't it better to specify something like InstanceNameSuffix? So instead of requiring the user to specify 'Sonarr', you just append it. like "{title} - Sonarr {InstanceName}"
> 
> I would suggest that is not as encompassing; prime example is people who would like "{title} - 4k Sonarr" and "{title} - Low Res Sonarr". Contained within the string makes more sense as it is more flexible

Flexibility isn't always good, after all, Sonarr is the brand name.
I'm not opposed to the current impl, but I think it's unnecessarily complex. On top of that users are gonna be confronted with a warning about it having to contain 'Sonarr'.
Let's see what markus thinks.

**Comment by markus101** [comment]
When we talked about being able to change the name before Qstick and I weren’t sure whether it should be a suffix or prefix made the most sense. From a branding perspective I think `Sonarr X` makes the most sense, but I wasn’t exactly sure how to communicate that clearly and haven’t thought about it much since.

A few thoughts:


- Either enforcing it at the beginning or applying it ourselves seems reasonable
- Applying it ourselves would also avoid the existing issue where currently if it’s set in the config file (bypassing the validation) it can be set to whatever
- We could go with a hybrid approach where it either starts with Sonarr or we append it
- Since it’s the app name I like the idea of the user including it, otherwise there is a gap where they don’t want to include Sonarr in the app name because we’re including it, but they want to include it because that’s how it shows up in syslogs
- From a branding POV the regex should be case sensitive


TLDR, I think I’m okay with the user setting it somewhere in the app name (this allows a bit of flexibility for the user to choose `4K Sonarr` vs `Sonarr 4K`, but maybe we change it to must be at the beginning or end specifically and require it to be title case.

I’d also take it a step further to inject it at the start if someone attempts to bypass the validation.

File: frontend/src/Settings/General/HostSettings.js [hunk 2]
```diff
@@ -71,6 +72,22 @@ function HostSettings(props) {
...
+        advancedSettings={advancedSettings}
+        isAdvanced={true}
+      >
+        <FormLabel>Instance Name</FormLabel>
+
+        <FormInputGroup
+          type={inputTypes.TEXT}
+          name="instanceName"
+          helpText="Instance name in tab and for Syslog app name"
+          helpTextWarning="Requires restart to take effect"
```
Comment (markus101) [inline frontend/src/Settings/General/HostSettings.js:85]: Make sure this prompts the user to restart, currently it doesn’t, but reload applies it immediately, which feels a bit weird.

**Comment by RobinDadswell** [comment]
I would agree with case sensitive; I will get that in now. I can also see the logic of having it either start with or end with it as opposed to being in the middle so happy to put that in. 

I would agree that I think the user should definitely have to put the name in themselves as it affects both the frontend and the syslog functionality. 

@Qstick do you have any issues with the change to the branding here outline by Markus as I will filter them downstream as well :D

**Comment by Qstick** [comment]
No issues from me, we knew about the gap in being able to edit direct in config and needing to patch that up. I'm fine with the other stuff, my biggest thing is just keeping the App name requirements. However we feel is the best approach for that, I'm down with.

File: src/NzbDrone.Core/Configuration/ConfigFileProvider.cs [hunk 3]
```diff
@@ -202,14 +206,33 @@ public string UrlBase
...
+            {
+                var instanceName = GetValue("InstanceName", BuildInfo.AppName);
+
+                if (instanceName.StartsWith(BuildInfo.AppName) || instanceName.EndsWith(BuildInfo.AppName) )
+                {
+                    return instanceName;
+                }
+                return BuildInfo.AppName;
+            }
+        }
```
Comment (RobinDadswell) [inline src/NzbDrone.Core/Configuration/ConfigFileProvider.cs:221]: @markus101 this is to stop it being worked around from not having Sonarr in the InstanceName
